app/models/concerns/labware_creators/donor_pooling_calculator.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+# This module contains calculations for donor pooling.
+module LabwareCreators::DonorPoolingCalculator
+  extend ActiveSupport::Concern
+
+  # Splits wells into groups by study and project. Wells are grouped
+  # based on the study and project of the first aliquot in each well. Returns
+  # an array of groups, where each group is an array of wells with the same
+  # study and project.
+  #
+  # @param group [Array<Well>] The group of wells to be split.
+  # @return [Array<Array<Well>>] An array of well groups.
+  def split_single_group_by_study_and_project(group)
+    group.group_by { |well| [well.aliquots.first.study.id, well.aliquots.first.project.id] }.values
+  end
+
+  # Splits groups ensuring unique donor_ids within each group. Iterates over
+  # each group, creating subgroups with wells from a unique donor. The first
+  # occurrences of unique donor_ids are grouped, then the second occurrences,
+  # and so on. This prevents combining samples with the same donor_id.
+  #
+  # @param groups [Array<Array<Well>>] Array of well groups to be split.
+  # @return [Array<Array<Well>>] Array of subgroups split by donor ID.
+  def split_groups_by_unique_donor_ids(groups)
+    groups.flat_map { |group| split_single_group_by_unique_donor_ids(group) }
+  end
+
+  # Splits a single group of wells by donor_ids. This method is used by the
+  # 'split_groups_by_unique_donor_ids' method. It iteratively segregates wells with
+  # the first encountered instance of each unique donor_id into a separate
+  # subgroup. This process continues until there are no wells left in the
+  # original group. The result is a collection of subgroups, each containing
+  # wells from distinct donors.
+  #
+  # @param group [Array<Well>] The group of wells to split.
+  # @return [Array<Array<Well>>] An array of subgroups, each containing wells
+  #   from different donors.
+  def split_single_group_by_unique_donor_ids(group)
+    group = group.dup
+    output = []
+    while group.any?
+      subgroup = []
+      unique_donor_ids(group).each do |donor_id|
+        index = group.index { |well| well.aliquots.first.sample.sample_metadata.donor_id == donor_id }
+        subgroup << group.delete_at(index) unless index.nil?
+      end
+      output << subgroup
+    end
+    output
+  end
```

Comment:
I think this works well. The `while` loop makes me a little nervous but hopefully no infinite loop possibilities here!

Reply:
I can't think of a way to achieve that but I have used a counter to address the doubt.